casepro/msgs/views.py

```diff
@@ -216,11 +233,62 @@ def get_context_data(self, **kwargs):
             return context
 
         def render_to_response(self, context, **response_kwargs):
+            results = []
+            for m in context['object_list']:
+                msg = m.as_json()
+
+                if context.get('user_id'):
+                    msg['lock'] = m.get_lock(context.get('user_id'))
+
+                results.append(msg)
+
             return JsonResponse({
-                'results': [m.as_json() for m in context['object_list']],
+                'results': results,
                 'has_more': context['has_more']
             }, encoder=JSONEncoder)
 
+    class Lock(OrgPermsMixin, SmartTemplateView):
+        """
+        AJAX endpoint for updating messages with a date and user id.
+        Takes a list of message ids.
+        """
+        @classmethod
+        def derive_url_pattern(cls, path, action):
+            return r'^message/lock/(?P<action>\w+)/$'
+
+        def post(self, request, *args, **kwargs):
+            org = request.org
+            user = request.user
+
+            action = kwargs['action']
+
+            message_ids = request.json['messages']
+            messages = org.incoming_messages.filter(org=org, backend_id__in=message_ids)
+
+            lock_messages = []
+
+            if action == 'lock':
+                for message in messages:
+                    if message.get_lock(request.user.id):
+                        lock_messages.append(message.backend_id)
+
+                if not lock_messages:
+                    for message in messages:
+                        message.locked_on = timezone.now()
+                        message.locked_by = user
+                        message.save()
+
+            elif action == 'unlock':
+                for message in messages:
+                    message.locked_on = None
+                    message.locked_by = None
+                    message.save()
```

Comment:
Any time you call `.save()` to update a few fields, you should specify those fields in the `update_fields` arg. Minimises chances of writing stale data